### ISSUE TITLE
10.12 compatibility?

### ISSUE THREAD

**Issue by sporkman** [pascalpp/FreeRuler#65]
First off, thanks for this! I was just wondering what would become of this app when my laptop gets upgraded to 10.15, and you're already here (before a few paid apps I'm waiting on even).

Second, should this be able to run on 10.12? I know I can use the old version, but was just curious if the compatibility was a "new at cocoa" issue or if the ruler is using some fancy features that aren't in 10.12.

**Comment by pascalpp** [comment]
@sporkman I'm actually not sure how far back it will run. I've only tested it on 10.14 Mojave, but I have confirmation that it works on 10.15 Catalina. If you're able to test it on 10.12 and report back, I'd be happy to hear how it goes. https://github.com/pascalpp/FreeRuler/releases

**Comment by sporkman** [comment]
Sorry - I meant to note that it's blocked from running on 10.12 - I get the "NO!" app overlay and this message on trying to run it. I think somewhere in the xcode project settings there are checkboxes for which versions of the OS you want to build for...
![screenshot-dt 2019-10-14 at 3 21 40 PM](https://user-images.githubusercontent.com/510750/66777269-97c0a280-ee96-11e9-860b-e9980e740f28.png)

**Comment by pascalpp** [comment]
Ah, okay. I'll have a look and see if I can expand support. Mind if I pass you some test builds here?

**Comment by sporkman** [comment]
Sure!

**Comment by pascalpp** [comment]
@sporkman lemme know if this works for you:
[free-ruler-2.0.2-b1.zip](https://github.com/pascalpp/FreeRuler/files/3726238/free-ruler-2.0.2-b1.zip)

**Comment by sporkman** [comment]
OK, on that one the restriction is gone, but it crashes on launch.

"Corpse notify", very seasonal!

Specifically:

"Library not loaded: @rpath/libswiftAppKit.dylib"

```
Process:               Free Ruler [78199]
Path:                  /Volumes/VOLUME/*/Free Ruler 2.app/Contents/MacOS/Free Ruler
Identifier:            com.pascal.freeruler
Version:               2.0.2-b1 (216)
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           Free Ruler [78199]
User ID:               502

Date/Time:             2019-10-14 16:08:15.344 -0400
OS Version:            Mac OS X 10.12.6 (16G2136)
Report Version:        12
Anonymous UUID:        6A2C75A7-0598-A9A8-6394-2B47C26919F5


Time Awake Since Boot: 770000 seconds

System Integrity Protection: disabled

Notes:                 Translocated Process

Crashed Thread:        0

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    DYLD, [0x1] Library missing

Application Specific Information:
dyld: launch, loading dependent libraries

Dyld Error Message:
  Library not loaded: @rpath/libswiftAppKit.dylib
  Referenced from: /Volumes/VOLUME/*/Free Ruler 2.app/Contents/MacOS/Free Ruler
  Reason: image not found

Binary Images:
       0x105f3e000 -        0x105f56ff7 +com.pascal.freeruler (2.0.2-b1 - 216) <10BBC831-4724-3989-A43B-EC3D98059C69> /var/folders/*/Free Ruler 2.app/Contents/MacOS/Free Ruler
       0x105f72000 -        0x105fafdc7  dyld (434) <33DB4E37-BC29-37A4-92AB-30328E66A8FA> /usr/lib/dyld
    0x7fff9db98000 -     0x7fff9e971ff3  com.apple.AppKit (6.9 - 1504.84.100) <E39CD613-0104-3E26-B424-BE83CED656F9> /System/Library/Frameworks/AppKit.framework/Versions/C/AppKit
    0x7fffa0062000 -     0x7fffa04fefe7  com.apple.CoreFoundation (6.9 - 1350.10) <60A73360-EE24-33A2-BF37-08CB8B96D2D9> /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
    0x7fffa04ff000 -     0x7fffa0b81fff  com.apple.CoreGraphics (2.0 - 1070.22.1) <8A053296-AA99-371C-B9C5-FCBF87C6CEFE> /System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics
    0x7fffa1add000 -     0x7fffa1e85ff7  com.apple.Foundation (6.9 - 1349.93) <A3CF1205-DE81-37FE-A574-0D381DECF122> /System/Library/Frameworks/Foundation.framework/Versions/C/Foundation
    0x7fffb42f2000 -     0x7fffb42f3ffb  libSystem.B.dylib (1238.60.2) <58358A17-4318-3AD9-B18E-CC36502B9127> /usr/lib/libSystem.B.dylib
    0x7fffb4fa0000 -     0x7fffb5372047  libobjc.A.dylib (709.1) <70614861-0340-32E2-85ED-FE65759CDFFA> /usr/lib/libobjc.A.dylib

Model: iMac17,1, BootROM 161.0.0.0.0, 4 processors, Intel Core i5, 3.19 GHz, 16 GB, SMC 2.33f10
Graphics: Intel HD Graphics 530, Intel HD Graphics 530, Built-In
...

```

**Comment by pascalpp** [comment]
Hmm odd. It looks like you might have renamed the app to 'Free Ruler 2'? That shouldn't matter, but if you did, can you try renaming it back to Free Ruler and see if it still runs?

**Comment by sporkman** [comment]
Same thing. Google suggests this is something in xcode settings and embedding needed libraries:

> To fix your problem try the following steps:
> 
> - go to your app target's Build Settings in Xcode
> - find Always Embed Swift Standard Libraries setting
> - set it from NO (which is default for Obj-C projects) to YES
> - After it clean the project (Shift+Cmd+K) and then build (Cmd+B) it again.

**Comment by pascalpp** [comment]
K thanks for that. Try this build:
[free-ruler-2.0.2-b2.zip](https://github.com/pascalpp/FreeRuler/files/3726374/free-ruler-2.0.2-b2.zip)

**Comment by sporkman** [comment]
Still failing. Another thing I found googling relates to the app referencing the bundled libraries (specifically Swift I think).

It notes that otool should show the system path as well as a path pointing to the bundled Swift lib, but I'm not seeing the latter:

```
frankentosh:Downloads spork$ otool -l Free\ Ruler.app/Contents/MacOS/Free\ Ruler |grep -B 1 -A 2 RPATH
Load command 23
          cmd LC_RPATH
      cmdsize 32
         path /usr/lib/swift (offset 12)
frankentosh:Downloads spork$
```
I'm looking at the tests in the highlighted answer here:

https://forums.developer.apple.com/thread/115551

Just stabbing in the dark here. I see the lib is bundled, but the app doesn't seem to know this...

Maybe I should try building.

**Comment by sporkman** [comment]
Oh, if I do this very bad and dangerous thing it launches!

```
frankentosh:Downloads spork$ sudo ln -s /Users/spork/Downloads/Free\ Ruler.app/Contents/Frameworks /usr/lib/swift
frankentosh:Downloads spork$ open Free\ Ruler.app/
frankentosh:Downloads spork$
```
![screenshot-dt 2019-10-14 at 5 33 10 PM](https://user-images.githubusercontent.com/510750/66784532-bdef3e00-eea8-11e9-9330-56bbc5f6c733.png)

(undoing that now)

So good news - the app can run on 10.12 (and probably 10.13, and maybe even earlier).  Bad news - no idea how to make it know it has the libraries already bundled!

**Comment by andrewkr** [comment]
I can't run Free Ruler 2.01 with 10.13.6.

![Screen Shot 2019-10-16 at 21 59 30](https://user-images.githubusercontent.com/588285/66913506-7d3a2500-f060-11e9-900c-e15992b39abf.png)

(Have I put this comment in the correct place?)

**Comment by andrewkr** [comment]
> @sporkman I'm actually not sure how far back it will run. I've only tested it on 10.14 Mojave, but I have confirmation that it works on 10.15 Catalina. If you're able to test it on 10.12 and report back, I'd be happy to hear how it goes. https://github.com/pascalpp/FreeRuler/releases

I've just tried it on my 10.15 Mac Mini and it isn't running. 
Application Not Responding
Force Quit is the option

**Comment by jeff-h** [comment]
@andrewkr wow that's really weird. It runs just fine on my 10.15. Could you try downloading it from the App store?

**Comment by sporkman** [comment]
@andrewkr - if the suggestion of running the official one (not the test version for older os-x versions posted in this issue) does not work, I'd open a new issue that specifically notes that you're seeing it fail on 10.15.

**Comment by andrewkr** [comment]
> @andrewkr - if the suggestion of running the official one (not the test version for older os-x versions posted in this issue) does not work, I'd open a new issue that specifically notes that you're seeing it fail on 10.15.

Certainly, but it is worse than that. It won't even download. :(

**Comment by pascalpp** [comment]
@andrewkr You're not able to download the app from the app store? Here's a direct link: https://apps.apple.com/us/app/free-ruler/id1483172210?ls=1&mt=12

**Comment by gnachman** [comment]
Sounds like you need to change the rpath at link time. Here's how I set it in my project, and I think it'll work for you too since you have the Swift libs in your Frameworks folder:

<img width="912" alt="Screen Shot 2019-10-17 at 9 56 06 PM" src="https://user-images.githubusercontent.com/427546/67066838-07e65700-f129-11e9-9e5a-21842577ae54.png">

**Comment by pascalpp** [comment]
@gnachman Thanks! I'll give that a try.

**Comment by pascalpp** [comment]
@sporkman Here's a build that hopefully has the Swift frameworks linked correctly. Can you try this on 10.12 and let me know if it works for you?

[free-ruler-2.0.2-b3.zip](https://github.com/pascalpp/FreeRuler/files/3742396/free-ruler-2.0.2-b3.zip)

**Comment by andrewkr** [comment]
> @andrewkr You're not able to download the app from the app store? Here's a direct link: https://apps.apple.com/us/app/free-ruler/id1483172210?ls=1&mt=12

@pascalpp  thanks for the direct link. It still wouldn't download. It kept getting stuck at, I think, 389 kb! 
Eventually, I removed from the applications directory the other version I had - 2.01 - instead of just renaming it. Again, it paused for a very long time at that 389 kb mark, and then finally the App Store screen showed "Open".

Sorry to have distracted you here with my issue.

**Comment by pascalpp** [comment]
@andrewkr can you try the [free-ruler-2.0.2-b3.zip](https://github.com/pascalpp/FreeRuler/files/3742396/free-ruler-2.0.2-b3.zip)
 build above on your 10.13 Mac?

**Comment by andrewkr** [comment]
> @andrewkr can you try the [free-ruler-2.0.2-b3.zip](https://github.com/pascalpp/FreeRuler/files/3742396/free-ruler-2.0.2-b3.zip) build on your 10.13 Mac?

I had seen it and started downloading it. 
And it works!

**Comment by pascalpp** [comment]
@andrewkr Nice. Does it run on your 10.15 Catalina Mac as well?

**Comment by andrewkr** [comment]
I was just playing with the options on the 10.13 Mac, and I noticed that when I select 'Float rulers', there's a tick ✓ next to it. It's the same for 'Group rulers'. But 'Show ruler shadow' doesn't show the tick. 
It would be good, in my mind, if there was a consistent approach.

I'll check the 2.03 version a bit later. I can't get to the machine just yet. (My time zone is GMT +11)

**Comment by andrewkr** [comment]
Sorry. I meant to include the screenshot illustrating what I was talking about.

![Screen Shot 2019-10-18 at 16 32 26](https://user-images.githubusercontent.com/588285/67068258-344fb100-f1c5-11e9-94ad-1c9bc166ecc4.png)

**Comment by pascalpp** [comment]
Good catch on the checkmark issue, not sure what happened there. Ticketed: https://github.com/pascalpp/FreeRuler/issues/69

**Comment by andrewkr** [comment]
I've just downloaded and run the 2.03 version (as updated) on my 10.15 Catalina Mac.

Yes, it does run.

That's an interesting change you made in #70. The 'tick' (as we call them in Australia and practically everywhere in the English-speaking world except the U.S.) is now permanently on.

**Comment by pascalpp** [comment]
@andrewkr Yeah the bug is that it only updates the menu item when you first launch the app. If you turn off the shadow and relaunch the app then the tick will be off. The fix in #70 is to observe that pref change and update the menu item every time it changes.

Okay I'm going to merge #71 cut a release for 2.02. Thanks for testing!

**Comment by sporkman** [comment]
I know you closed, but wanted to confirm that this works on 10.12.

And thanks @gnachman!

![screenshot-dt 2019-10-18 at 3 30 22 PM](https://user-images.githubusercontent.com/510750/67122681-56910100-f1bc-11e9-941f-1fb939f99612.png)
